src/extension_commands.ml

```diff
@@ -534,6 +534,212 @@ module Copy_type_under_cursor = struct
+    | _ -> false
+
+  let ocaml_lsp_doesnt_support_search_by_type ocaml_lsp =
+    not (Ocaml_lsp.can_handle_search_by_type ocaml_lsp)
+
+  let get_search_results ~query ~limit ~with_doc ~position text_editor client =
+    let doc = TextEditor.document text_editor in
+    let uri = TextDocument.uri doc in
+    Custom_requests.(
+      send_request
+        client
+        Type_search.request
+        (Type_search.make ~uri ~position ~limit ~query ~with_doc ()))
+    |> Promise.catch ~rejected:(fun _ -> Promise.return [])
+
+  let input_box =
+    (* Re-using the same instance of the input box allows us to remember the
+       last input. *)
+    let box =
+      InputBox.set
+        (Window.createInputBox ())
+        ~title:"Search By Type"
+        ~ignoreFocusOut:false
+        ~placeholder:"int -> string / -int +string"
+        ~prompt:
+          "Perform a search by type request by providing a type signature to \
+           look for"
+        ()
+    in
+    let _disposable =
+      InputBox.onDidChangeValue
+        box
+        ~listener:(fun _ -> InputBox.set_validationMessage box None)
+        ()
+    in
+    box
+
+  let rec display_search_results query results text_editor position client =
+    let quickPickItems =
+      List.map
+        results
+        ~f:(fun (res : Custom_requests.Type_search.type_search_result) ->
+          QuickPickItem.create
+            ~label:res.name
+            ~description:res.typ
+            ~detail:(Option.value ~default:"" res.doc)
+            ())
+    in
+    let module QuickPick = Vscode.QuickPick.Make (QuickPickItem) in
+    let quickPick =
+      QuickPick.set
+        (Window.createQuickPick (module QuickPickItem) ())
+        ~title:"Type/Polarity Search Results"
+        ~activeItems:[]
+        ~busy:false
+        ~enabled:true
+        ~placeholder:"Select an item to insert it to the editor"
+        ~selectedItems:[]
+        ~ignoreFocusOut:false
+        ~items:quickPickItems
+        ~buttons:[ Window.quickInputButtonBack ]
+        ()
+    in
+    let _disposable =
+      QuickPick.onDidTriggerButton
+        quickPick
+        ~listener:(fun _ -> show_query_input text_editor client)
+        ()
+    in
+    let _disposable =
+      QuickPick.onDidAccept
+        quickPick
+        ~listener:(fun () ->
+          match QuickPick.selectedItems quickPick with
+          | Some (item :: _) ->
+            let value = QuickPickItem.label item in
+            let _ =
+              Vscode.TextEditor.edit
+                text_editor
+                ~callback:(fun ~editBuilder ->
+                  Vscode.TextEditorEdit.insert
+                    editBuilder
+                    ~location:position
+                    ~value)
+                ()
+            in
+            QuickPick.hide quickPick
+          | _ ->
+            display_search_results query results text_editor position client)
+        ()
+    in
```

Comment:
I think we should also `dispose` the quickpick when `onDidHide` to be on the safe side.